Add student and teacher links to surveys

willgearty opened the pull request:
I added links to the student and teacher surveys (and the teacher survey review page) to the survey management page (if there are multiple surveys for a single category, we specify the survey ID for each link). I also changed the `isStep` behavior for the survey module such that it will be included in the registration steps if 1) the `survey_isstep` tag is set to True and 2) the first class block of the program has started.

Fixes https://github.com/learning-unlimited/ESP-Website/issues/3035.

kkbrum commented on esp/esp/tagdict/__init__.py:
```diff
@@ -136,6 +136,7 @@
     'switch_time_program_attendance': (False, 'At what time should the student onsite webapp use program attendance if available (instead of enrollment) to determine if a class is full? If blank, program attendance numbers will not be used. Format: HH:MM where HH is in 24 hour time.', None, 'learn', True),
     'switch_lag_class_attendance': (False, 'How many minutes into a class should the student onsite webapp use class attendance numbers if available (instead of enrollment or program attendance) to determine if a class is full? If blank, class attendance numbers will not be used.', None, 'learn', True),
     'student_lottery_group_max': (False, 'What is the maximum number of students that can be in the same lottery group? (set to 1 to not allow groups)', '4', 'learn', True),
+    'survey_isstep': (True, 'Should the student and teacher surveys be shown as steps in student and teacher registration respectively once the event has started?', False, 'manage', True),
```
Why is the survey_isstep a program tag while webapp_isstep is a global tag? I think these isstep tags should be consistent, otherwise it makes them hard to find as an admin. I think as long as we import tags when we import a program, I would probably prefer both to be program tags.

kkbrum replied:
Looking at the code changes I don't have any clue as to how to do this, but ideally even with the "survey_isstep", the step should only show up if a student/teacher survey exists for the program. For example if "isstep" is true and there's actually only a teacher survey, it would be not great for the student survey link to show up in student registration and have this error pop up:
![image](https://user-images.githubusercontent.com/25392075/88609134-6134cc80-d051-11ea-8850-c089f51771d9.png)

kkbrum replied:
The step in student registration for "surveys" shows up smack in the middle of the list. I think it should probably be the last item. Could we update the seq for the module to make it lower on the list?

kkbrum commented on esp/templates/program/modules/surveymanagement/main.html:
```diff
@@ -16,16 +16,24 @@ <h1>Surveys for {{ program.niceName }}</h1>
+        <li><a href="/teach/{{ program.getUrlBase }}/survey{% if counts.teach > 1 %}?survey_id={{ s.id }}{% endif %}">Link for teachers</a></li>
+    {% elif s.category == "learn" %}
+        <li><a href="/learn/{{ program.getUrlBase }}/survey{% if counts.learn > 1 %}?survey_id={{ s.id }}{% endif %}">Link for students</a></li>
+        <li><a href="/teach/{{ program.getUrlBase }}/survey/review{% if counts.learn > 1 %}?survey_id={{ s.id }}{% endif %}">Link for teachers to review results</a></li>
```
Perhaps this link should also get added to teacher registration (if student surveys exist)

willgearty commented:
OK, I think that should address all of your comments @kkbrum!

First, I've split the survey program module into two separate modules, one for student registration and one for teacher registration (a la #2125). I made a migration that hopefully changes existing modules such that programs with the surveys modules will keep those modules and will just use the new handlers (although I had a hard time testing this, so please let me know if it works in review). Ultimately, nothing should behave differently on the frontend with this change, but separating the two modules to separate handlers does allow us to have separate `isStep()` functions, which also means we can have separate isStep tags (#3074). I'll make separate tags for the webapps in a separate PR.

Second, I've added a link for teachers to go to their survey review page. This required a little bit more work than just adding a link because we want to make sure we show the survey page (and not an error page) in cases where there is no teacher survey but there are student survey results for the teacher to review.

Finally, I fixed a bug that I discovered where we weren't linking to the correct page for section or general surveys if there were multiple surveys of the same category for a single program.

kkbrum commented on esp/templates/survey/survey_form.html:
```diff
@@ -22,20 +22,24 @@
         <p class="message">This survey does not have any questions.</p>
     {% endif %}
+    {% if student_results %}
+        <p class="message">You can view the results of the student survey(s) for your class(es) by <a href="{{ request.path }}/review">going here</a>.
```
If there are multiple teacher surveys, then this shows up only after you click on one of the teacher surveys, which is weird to me since the student survey is not connected to the teacher surveys. So at the very least I think it should show up on the page with the list of multiple surveys instead of when you click one of them. 

However, I've realized since my original review that I would prefer if the survey results were linked in the "Manage your classes" boxes at the bottom of the techerreg page instead. Since there's a box per class and that box should contain everything about the class, I think that's the most logical place for a teacher to find their survey results. I think I would just add a button "View survey results" under the "Email ___ Students" button (if survey results exist).

kkbrum replied:
(you can change that last condition to be just if a student survey exists. If there are no results I'm okay with having the link be there and say "no results yet! check back later" when clicked)

kkbrum commented on esp/esp/program/modules/migrations/0023_auto_20200730_2006.py:
```diff
@@ -0,0 +1,80 @@
+
+from django.db import migrations
+
+def set_my_defaults(apps, schema_editor):
```
this worked for me!

kkbrum commented on esp/esp/program/modules/handlers/teachersurveymodule.py:
```diff
@@ -78,17 +61,16 @@ def teachers(self, QObject = False):
-        return False
+        return (Tag.getBooleanTag('teacher_survey_isstep', program=self.program, default=False) and
+                self.program.getTimeSlots()[0].start < datetime.datetime.now() and
+                self.program.getSurveys().filter(category__in = ["learn","teach"]).exists())
```
Just a reminder that if you implement my comment about moving the survey results link, then this should just check for a teacher survey

kkbrum commented on esp/esp/program/modules/handlers/teachersurveymodule.py:
```diff
@@ -78,17 +61,16 @@ def teachers(self, QObject = False):
     def isStep(self):
-        return False
+        return (Tag.getBooleanTag('teacher_survey_isstep', program=self.program, default=False) and
+                self.program.getTimeSlots()[0].start < datetime.datetime.now() and
```
I checked this for you and it works as desired!

willgearty commented:
OK, I added the links as requested to the main teacher registration page (each link is specific to a class). I also added the survey review link to the "Choose a Survey" page if there are multiple teacher surveys. I figured it couldn't hurt to keep it on the original page, so I didn't change the isStep condition. Thanks as always for the great feedback @kkbrum!
